cpp/examples/tutorial_examples/dataset_example.cc

```diff
@@ -0,0 +1,169 @@
+    ARROW_ASSIGN_OR_RAISE(auto src_table, CreateTable())
+    std::shared_ptr<arrow::fs::FileSystem> setup_fs;
+    //Note this operates in the directory the executable is built in.
+    char setup_path[256];
+    getcwd(setup_path, 256);
+    ARROW_ASSIGN_OR_RAISE(setup_fs, arrow::fs::FileSystemFromUriOrPath(setup_path));
+    ARROW_ASSIGN_OR_RAISE(auto dset_path, CreateExampleParquetDataset(setup_fs,
+                                                                      ""));
+
+    //First, we need a filesystem object, which lets us interact with our local
+    //filesystem starting at a given path. For the sake of simplicity, that'll be
+    //the current directory.
+    std::shared_ptr<arrow::fs::FileSystem> fs;
+    //This feels pretty bad, but I wasn't finding great solutions that're
+    //system-generic -- could use some advice on how to set this up.
+    char init_path[256];
+    getcwd(init_path, 256);
+    ARROW_ASSIGN_OR_RAISE(fs, arrow::fs::FileSystemFromUriOrPath(init_path));
+
+    //A file selector lets us actually traverse a multi-file dataset.
+    arrow::fs::FileSelector selector;
+    selector.base_dir = "parquet_dataset";
+    //Recursive is a safe bet if you don't know the nesting of your dataset.
+    selector.recursive = true;
+    //Making an options object lets us configure our dataset reading.
+    arrow::dataset::FileSystemFactoryOptions options;
+    // We'll use Hive-style partitioning. We'll let Arrow Datasets infer the partition
+    // schema. We won't set any other options, defaults are fine.
+    options.partitioning = arrow::dataset::HivePartitioning::MakeFactory();
+    auto read_format =
+            std::make_shared<arrow::dataset::ParquetFileFormat>();
+    //Now, we get a factory that will let us get our dataset -- we don't have the
+    //dataset yet!
+    auto factory = arrow::dataset::FileSystemDatasetFactory::Make(fs,
+                                                                  selector,
+                                                                  read_format,
+                                                                  options)
+            .ValueOrDie();
+    //Now we read into our dataset from the factory.
+    auto read_dataset = factory->Finish().ValueOrDie();
+    // Print out the fragments
+    for (const auto& fragment : read_dataset->GetFragments().ValueOrDie()) {
+        std::cout << "Found fragment: " << (*fragment)->ToString() << std::endl;
+    }
+
+    //Scan dataset into a Table -- once this is done, you can do
+    //normal table things with it, like computation and printing. However, now you're
+    //also dedicated to being in memory.
+    auto read_scan_builder = read_dataset->NewScan().ValueOrDie();
+    auto read_scanner = read_scan_builder->Finish().ValueOrDie();
+    std::shared_ptr<arrow::Table> table = read_scanner->ToTable().ValueOrDie();
+    std::cout << table->ToString();
+
+    //Now, let's get a table out as a dataset!
```

Comment:
So it turns out that `ScanBatches()` does not get a `RecordBatchReader` -- instead, `ToRecordBatchReader()` does. Do we want to have `ToRecordBatchReader()` and `FromRecordBatchReader()` back-to-back? It's a lot more code (especially when avoiding `ValueOrDie()`), and fairly awkward to read.

Reply:
Yeah, it is awkward to go from Scanner -> RBR -> Scanner. Maybe for the sake of this intro just directly pass the scanner over to the write function?

TBH I'm very thrown off by the fact that we pass a scanner to write a dataset; I didn't expect it to be there. Also am anticipating Scanner to [disappear sometime soon](https://issues.apache.org/jira/browse/ARROW-16409).

Reply:
I think the pause at the Table is valuable to have before we introduce Acero, because before Acero, there's not much to do if we can't show off "here's the Table you know and love."

Reply:
How about `Dataset -> Scanner -> Table -> RecordBatchReader -> Scanner -> WriteDataset` then? Instead of `Dataset -> Scanner -> Table -> InMemoryDataset -> Scanner -> WriteDataset`?

Reply:
So a `TableBatchReader`, in particular?

Reply:
At this point we don't actually "scan" the scanner.  We simply need it for the options :grimacing:  It would be pretty easy (and rather useful) to change:

```
Status FileSystemDataset::Write(const FileSystemDatasetWriteOptions& write_options,
                                std::shared_ptr<Scanner> scanner)
```

to:

```
Status FileSystemDataset::Write(const FileSystemDatasetWriteOptions& write_options, std::shared_ptr<ScanOptions> options, std::shared_ptr<Dataset> dataset)
```

Even better, we should create (untested but should work):

```
Status FileSystemDataset::Write(const FileSystemDatasetWriteOptions& write_options,
                                std::shared_ptr<Table> table) {
  std::shared_ptr<compute::ExecContext> exec_context =
      std::make_shared<compute::ExecContext>();

  ARROW_ASSIGN_OR_RAISE(auto plan, compute::ExecPlan::Make(exec_context.get()));

  RETURN_NOT_OK(
      compute::Declaration::Sequence(
          {
              {"table_source", compute::TableSourceNodeOptions{std::move(table)}},
              {"write", WriteNodeOptions{write_options}},
          })
          .AddToPlan(plan.get()));

  RETURN_NOT_OK(plan->StartProducing());
  return plan->finished().status();
}
```

Creating a dataset from a scanner (or rather, scan_options) makes a lot more sense when you consider the fact that the source might be a larger-than-memory dataset (this is very common when re-partitioning a dataset, or converting a dataset from CSV to parquet for example).

However, we definitely do not need to be involving something as complex as a scan node to write an in-memory table to a dataset.

Reply:
Could you turn this into a JIRA issue?

Reply:
@wjones127 I've swapped out the InMemoryDataset for a TableBatchReader. Look good?